src/util/io.rs

```diff
@@ -63,11 +68,105 @@ impl Read for FileChunk {
   }
 }
 
+/// Positional write trait.
+///
+/// Tracks the current position in the stream, should be used together with `Write`.
+/// This does not require &mut self to get the position.
+pub trait Position {
+  /// Returns the current position in the stream.
+  fn pos(&self) -> u64;
+}
+
+/// Struct that represents `File` output stream with position tracking.
+/// Used as a sink in file writer.
+pub struct FileSink {
+  buf: BufWriter<File>,
+  // This is not necessarily position in the underlying file,
+  // but rather current position in the sink.
+  pos: u64
+}
+
+impl FileSink {
+  /// Creates new file sink.
+  /// Position is set to whatever position file has.
+  pub fn new(file: &File) -> Self {
+    let mut owned_file = file.try_clone().unwrap();
+    let pos = owned_file.seek(SeekFrom::Current(0)).unwrap();
+    Self {
+      buf: BufWriter::new(owned_file),
+      pos: pos
+    }
+  }
+}
+
+impl Write for FileSink {
+  fn write(&mut self, buf: &[u8]) -> Result<usize> {
+    let num_bytes = self.buf.write(buf)?;
+    self.pos += num_bytes as u64;
+    Ok(num_bytes)
+  }
+
+  fn flush(&mut self) -> Result<()> {
+    self.buf.flush()
+  }
+}
+
+impl Position for FileSink {
+  fn pos(&self) -> u64 {
+    self.pos
+  }
+}
+
+// ----------------------------------------------------------------------
+// Thrift IO.
+
+/// A thin read-only wrapper on `T: Read` to be used by Thrift transport.
+pub struct TInputBuffer<'a, T> where T: 'a + Read {
```

Comment:
These (both `TInputBuffer` and `TOutputBuffer` do not seem necessary anymore as we can pass `Read` and `Write` to Thrift transport.

Reply:
Yes, you are right. I also made changes in the prototype branch to check it compiles. Will remove.